cql3/expr/expression.hh

```diff
@@ -72,13 +72,29 @@ class unresolved_identifier;
 using expression = std::variant<bool, conjunction, binary_operator, column_value, column_value_tuple, token,
                                 unresolved_identifier>;
 
+template <typename T, typename Variant = expression>
+struct is_expression_element : std::false_type {
+};
+
+template <typename T, typename... Elements>
+struct is_expression_element<T, std::variant<T, Elements...>> : std::true_type {
+};
+
+template <typename T, typename U, typename... Elements>
+struct is_expression_element<T, std::variant<U, Elements...>> : is_expression_element<T, std::variant<Elements...>>{
+};
+
+template <typename T>
+concept expression_element = is_expression_element<T>::value;
+
 // An expression variant element can't contain an expression by value, since the size of the variant
 // will be infinite. `nested_expression` contains an expression indirectly, but has value semantics and
 // is copyable.
 class nested_expression {
     std::unique_ptr<expression> _e;
 public:
     nested_expression(expression e);
+    nested_expression(expression_element auto e);
```

Comment:
Just to make sure I understand: is `expression_element auto e` syntactic sugar for "this is actually a template parameter restricted by the `expression_element` concept"?

Reply:
Yes, see https://en.cppreference.com/w/cpp/language/function_template#Abbreviated_function_template